toolchain/check/function.cpp

```diff
@@ -73,49 +73,29 @@ auto CheckFunctionTypeMatches(Context& context,
 auto CheckFunctionReturnType(Context& context, SemIRLoc loc,
                              SemIR::Function& function,
                              SemIR::GenericInstanceId specific_id)
-    -> SemIR::Function::ReturnSlot {
-  // If we have already checked the return type, we have nothing to do.
-  if (function.return_slot != SemIR::Function::ReturnSlot::NotComputed &&
-      !specific_id.is_valid()) {
-    return function.return_slot;
-  }
-
-  if (!function.return_storage_id.is_valid()) {
-    // Implicit `-> ()` has no return slot.
-    function.return_slot = SemIR::Function::ReturnSlot::Absent;
-    return function.return_slot;
-  }
+    -> SemIR::Function::ReturnInfo {
+  auto return_info = function.GetReturnInfo(context.sem_ir(), specific_id);
 
-  auto return_type_id =
-      function.GetDeclaredReturnType(context.sem_ir(), specific_id);
-  CARBON_CHECK(return_type_id.is_valid())
-      << "Have return storage but no return type.";
+  // If we couldn't determine the return information due to the return type
+  // being incomplete, try to complete it now.
+  if (return_info.slot == SemIR::Function::ReturnSlot::Incomplete) {
```

Comment:
Had you considered restructuring this for early returns, in order to trim nested scopes? e.g.:

```suggestion
  if (return_info.slot.is_valid()) {
    return return_info;
  }
```

If you do this, might want to adjust comments appropriately. Also, line 94 could return directly.

But either way, probably use `is_valid` here if you're adding it.

Reply:
I think that structure would be awkward if there's ever a second reason why we can't use a type as a return type. Which there is: we can't use abstract class types as return types either. I should add a diagnostic for that... will send follow-up PR.